README.md
# colander

[![Build Status](https://travis-ci.org/Supermercato24/colander.svg?branch=master)](https://travis-ci.org/Supermercato24/colander) [![Go Report Card](https://goreportcard.com/badge/github.com/Supermercato24/colander)](https://goreportcard.com/report/github.com/Supermercato24/colander)

Aggregate multiple logs into single one by timestamp.


CHANGELOG.md
v0.2.0-public / 2018-01-09
==================

 * (build) repo is public
 * (build) minor fixes

v0.1.0-go-go / 2017-12-10
==================

 * (build) compiled with 1.9.2
 * (configs) try to grab gdo configuration
 * (libs) resend message to Queue 1 time


file/log_test.go
package file

import (
	"os"
	"path/filepath"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/supermercato24/colander/config"
	"github.com/supermercato24/configuration"
)

const (
	numberOfLines   = 18
	dailyLogFile0   = "d2-2017-12-11.log"
	dailyLogFile1   = "d2-2017-12-11-ws01.log"
	dailyLogFile2   = "d2-2017-12-11-ws02.log"
	dailyLogFile3   = "day-2017-12-11.log"
	monthlyLogFile0 = "month-2017-12.log"
)

var (
	dailyLogBody0 = []byte(`[2017-12-01 23:16:20] d2-201712-11
[2017-12-01 23:21:00] d2-201712-11
[2017-12-01 23:23:40] d2-201712-11
[2017-12-01 23:23:50] d2-201712-11
[2017-12-01 23:24:20] d2-201712-11
`)
	dailyLogBody1 = []byte(`[2017-12-01 23:16:21] d2-201712-10-ws01
[2017-12-01 23:21:01] d2-201712-11-ws01
[2017-12-01 23:23:41] d2-201712-11-ws01
[2017-12-01 23:23:51] d2-201712-11-ws01
[2017-12-01 23:24:21] d2-201712-11-ws01
[2017-12-01 23:25:21] d2-201712-10-ws01
`)
	dailyLogBody2 = []byte(`[2017-12-01 23:16:22] d2-201712-10-ws02
[2017-12-01 23:21:02] d2-201712-11-ws02
[2017-12-01 23:23:42] d2-201712-11-ws02
[2017-12-01 23:23:52] d2-201712-11-ws02
[2017-12-01 23:24:22] d2-201712-11-ws02
[2017-12-01 23:25:22] d2-201712-10-ws02
[2017-12-01 23:26:22] d2-201712-10-ws01
`)
)

func TestDetectTimestamp(t *testing.T) {
	var validTimestamp []byte
	var expectedTimestamp time.Time
	var result time.Time

	validTimestamp = []byte("[2017-12-01 23:50:07]")
	expectedTimestamp = time.Date(2017, time.December, 1, 23, 50, 07, 0, time.UTC)
	result = detectTimestamp(validTimestamp)
	assert.Exactly(t, expectedTimestamp, result)

	validTimestamp = []byte("foo2017-12-01 23:50:07bar")
	expectedTimestamp = time.Date(2017, time.December, 1, 23, 50, 07, 0, time.UTC)
	result = detectTimestamp(validTimestamp)
	assert.Exactly(t, expectedTimestamp, result)

	validTimestamp = []byte("foo2017-12-01bar23:50:")
	expectedTimestamp = time.Time{}
	result = detectTimestamp(validTimestamp)
	assert.Exactly(t, expectedTimestamp, result)

	validTimestamp = []byte("foo2017-12-01bar23:50: 2017-12-01 23:50:07")
	expectedTimestamp = time.Date(2017, time.December, 1, 23, 50, 07, 0, time.UTC)
	result = detectTimestamp(validTimestamp)
	assert.Exactly(t, expectedTimestamp, result)
}

func logSetUp(t *testing.T, logNumber int) {
	var dailyLogFile string
	var dailyLogBody []byte

	switch logNumber {
	case 0:
		dailyLogFile = dailyLogFile0
		dailyLogBody = dailyLogBody0
	case 1:
		dailyLogFile = dailyLogFile1
		dailyLogBody = dailyLogBody1
	case 2:
		dailyLogFile = dailyLogFile2
		dailyLogBody = dailyLogBody2
	case 3:
		dailyLogFile = dailyLogFile3
		dailyLogBody = dailyLogBody2
	case 4:
		dailyLogFile = monthlyLogFile0
		dailyLogBody = dailyLogBody2
	}
	assert.NotEmpty(t, dailyLogFile)
	assert.NotEmpty(t, dailyLogBody)

	outputFile := filepath.Join(configuration.DirBinStorage, configuration.PathLogs, dailyLogFile)
	os.Remove(outputFile)
	fd, err := os.OpenFile(outputFile, os.O_CREATE|os.O_WRONLY, 0644)
	assert.NoError(t, err)

	//_, err = fd.Seek(int64(64), 0)
	//assert.NoError(t, err)

	n, err := fd.Write(dailyLogBody)
	assert.NoError(t, err)
	assert.NotZero(t, n)

	err = fd.Close()
	assert.NoError(t, err)
}

func logReadline(t *testing.T) {
	Glob(filepath.Join(configuration.DirBinStorage, configuration.PathLogs), "", func(matches *GlobMatches) {
		match := matches.Files[dailyLog1]
		assert.Exactly(t, dailyLog1, match.Category)

		for _, match := range match.Logs {
			if match.Day != 11 {
				continue
			}

			assert.Exactly(t, int64(11), match.Day)
			logs := LogReadLines(match.Paths)
			assert.NotEmpty(t, logs)

			assert.Len(t, logs, numberOfLines)

			break
		}
	})

	Glob(filepath.Join(configuration.DirBinStorage, configuration.PathLogs), "*2017-12-11*.log", func(matches *GlobMatches) {
		match := matches.Files[dailyLog1]
		assert.Exactly(t, dailyLog1, match.Category)

		for _, match := range match.Logs {
			assert.Exactly(t, int64(11), match.Day)
			logs := LogReadLines(match.Paths)
			assert.NotEmpty(t, logs)

			assert.Len(t, logs, numberOfLines)
		}
	})
}

func TestLog(t *testing.T) {
	assert.Exactly(t, config.Name, config.Name, "load init")

	assert.True(t, t.Run("setUp0", func(t *testing.T) { logSetUp(t, 0) }))
	assert.True(t, t.Run("setUp1", func(t *testing.T) { logSetUp(t, 1) }))
	assert.True(t, t.Run("setUp2", func(t *testing.T) { logSetUp(t, 2) }))

	assert.True(t, t.Run("logReadline", logReadline))
}


file/glob_test.go
package file

import (
	"path/filepath"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/supermercato24/colander/config"
	"github.com/supermercato24/configuration"
)

const (
	dailyLog0   = "day"
	dailyLog1   = "d2"
	monthlyLog0 = "month"
)

func globLogs(t *testing.T) {
	var day, d2, month bool

	Glob(filepath.Join(configuration.DirBinStorage, configuration.PathLogs), "", func(matches *GlobMatches) {
		for filesKey, match := range matches.Files {
			if filesKey == dailyLog0 {
				assert.Exactly(t, dailyLog0, filesKey)
				day = true
			} else if filesKey == dailyLog1 {
				assert.Exactly(t, dailyLog1, filesKey)
				d2 = true
			} else {
				assert.Exactly(t, monthlyLog0, filesKey)
				month = true
			}
			assert.Exactly(t, match.Category, filesKey)

			for logsKey, match := range match.Logs {
				assert.Exactly(t, int64(2017), match.Year)
				if filesKey == monthlyLog0 {
					assert.Exactly(t, int64(0), match.Day)
				} else {
					assert.Exactly(t, int64(12), match.Month)
				}
				assert.IsType(t, time.Time{}, logsKey)
				assert.NotZero(t, len(match.Paths))

				for _, match := range match.Paths {
					assert.True(t, filepath.IsAbs(match))
				}
			}
		}
	})

	assert.True(t, day)
	assert.True(t, d2)
	assert.True(t, month)
}

func TestGlob2(t *testing.T) {
	assert.Exactly(t, config.Name, config.Name, "load init")

	assert.True(t, t.Run("setUp0", func(t *testing.T) { logSetUp(t, 0) }))
	assert.True(t, t.Run("setUp3", func(t *testing.T) { logSetUp(t, 3) }))
	assert.True(t, t.Run("setUp4", func(t *testing.T) { logSetUp(t, 4) }))

	assert.True(t, t.Run("globLogs", globLogs))
}


file/sort_test.go
package file

import (
	"bufio"
	"os"
	"path/filepath"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/supermercato24/colander/config"
	"github.com/supermercato24/configuration"
)

func sortByTimestampWithoutPattern(t *testing.T) {
	Glob(filepath.Join(configuration.DirBinStorage, configuration.PathLogs), "", func(matches *GlobMatches) {
		match := matches.Files[dailyLog1]
		assert.Exactly(t, dailyLog1, match.Category)

		for _, match := range match.Logs {
			if match.Day != 11 {
				continue
			}

			assert.Exactly(t, int64(11), match.Day)
			logs := LogReadLines(match.Paths)
			assert.NotEmpty(t, logs)

			assert.Len(t, logs, numberOfLines)
			timestampTouchpoint0 := logs[0].Timestamp
			timestampTouchpoint1 := logs[numberOfLines/2-1].Timestamp
			timestampTouchpoint2 := logs[numberOfLines/3-1].Timestamp
			timestampTouchpoint3 := logs[numberOfLines/1-1].Timestamp

			sortedLogs := SortByTimestamp(logs)

			assert.Len(t, sortedLogs, numberOfLines)
			sortedTimestampTouchpoint0 := sortedLogs[0].Timestamp
			sortedTimestampTouchpoint1 := sortedLogs[numberOfLines/2-1].Timestamp
			sortedTimestampTouchpoint2 := sortedLogs[numberOfLines/3-1].Timestamp
			sortedTimestampTouchpoint3 := sortedLogs[numberOfLines/1-1].Timestamp
			assert.True(t,
				(timestampTouchpoint0.Unix() != sortedTimestampTouchpoint0.Unix()) ||
					(timestampTouchpoint1.Unix() != sortedTimestampTouchpoint1.Unix()) ||
					(timestampTouchpoint2.Unix() != sortedTimestampTouchpoint2.Unix()) ||
					(timestampTouchpoint3.Unix() != sortedTimestampTouchpoint3.Unix()))

			assert.Exactly(t, &sortedLogs, &logs, "reference")
			assert.True(t, (&sortedTimestampTouchpoint0 != &timestampTouchpoint0) ||
				(&sortedTimestampTouchpoint1 != &timestampTouchpoint1) ||
				(&sortedTimestampTouchpoint2 != &timestampTouchpoint2) ||
				(&sortedTimestampTouchpoint3 != &timestampTouchpoint3))

			resultPath := filepath.Join(configuration.DirBinStorage, configuration.PathLogs, "test_result.log")
			err := LogWriteLines(resultPath, sortedLogs)
			assert.NoError(t, err)

			fd, err := os.Open(resultPath)
			assert.NoError(t, err)

			ops := 0
			reader := bufio.NewReader(fd)

			for {
				line, _, err := reader.ReadLine()
				if err != nil {
					break
				}
				ops += 1
				assert.NoError(t, err)

				if ops == 1 {
					assert.Exactly(t, detectTimestamp(line), sortedTimestampTouchpoint0)
				}
			}
			assert.Exactly(t, ops, numberOfLines)
			fd.Close()

			break
		}
	})
}

func sortByTimestampWithPattern(t *testing.T) {
	Glob(filepath.Join(configuration.DirBinStorage, configuration.PathLogs), "*2017-12-11*.log", func(matches *GlobMatches) {
		match := matches.Files[dailyLog1]
		assert.Exactly(t, dailyLog1, match.Category)

		for _, match := range match.Logs {
			assert.Exactly(t, int64(11), match.Day)
			logs := LogReadLines(match.Paths)
			assert.NotEmpty(t, logs)

			assert.Len(t, logs, numberOfLines)
			timestampTouchpoint0 := logs[0].Timestamp
			timestampTouchpoint1 := logs[numberOfLines/2-1].Timestamp
			timestampTouchpoint2 := logs[numberOfLines/3-1].Timestamp
			timestampTouchpoint3 := logs[numberOfLines/1-1].Timestamp

			sortedLogs := SortByTimestamp(logs)

			assert.Len(t, sortedLogs, numberOfLines)
			sortedTimestampTouchpoint0 := sortedLogs[0].Timestamp
			sortedTimestampTouchpoint1 := sortedLogs[numberOfLines/2-1].Timestamp
			sortedTimestampTouchpoint2 := sortedLogs[numberOfLines/3-1].Timestamp
			sortedTimestampTouchpoint3 := sortedLogs[numberOfLines/1-1].Timestamp
			assert.True(t,
				(timestampTouchpoint0.Unix() != sortedTimestampTouchpoint0.Unix()) ||
					(timestampTouchpoint1.Unix() != sortedTimestampTouchpoint1.Unix()) ||
					(timestampTouchpoint2.Unix() != sortedTimestampTouchpoint2.Unix()) ||
					(timestampTouchpoint3.Unix() != sortedTimestampTouchpoint3.Unix()))

			assert.Exactly(t, &sortedLogs, &logs, "reference")
			assert.True(t, (&sortedTimestampTouchpoint0 != &timestampTouchpoint0) ||
				(&sortedTimestampTouchpoint1 != &timestampTouchpoint1) ||
				(&sortedTimestampTouchpoint2 != &timestampTouchpoint2) ||
				(&sortedTimestampTouchpoint3 != &timestampTouchpoint3))

			resultPath := filepath.Join(configuration.DirBinStorage, configuration.PathLogs, "test_result.log")
			err := LogWriteLines(resultPath, sortedLogs)
			assert.NoError(t, err)

			fd, err := os.Open(resultPath)
			assert.NoError(t, err)

			ops := 0
			reader := bufio.NewReader(fd)

			for {
				line, _, err := reader.ReadLine()
				if err != nil {
					break
				}
				ops += 1
				assert.NoError(t, err)

				if ops == 1 {
					assert.Exactly(t, detectTimestamp(line), sortedTimestampTouchpoint0)
				}
			}
			assert.Exactly(t, ops, numberOfLines)
			fd.Close()
		}
	})
}

func TestSort(t *testing.T) {
	assert.Exactly(t, config.Name, config.Name, "load init")

	assert.True(t, t.Run("setUp0", func(t *testing.T) { logSetUp(t, 0) }))
	assert.True(t, t.Run("setUp1", func(t *testing.T) { logSetUp(t, 1) }))
	assert.True(t, t.Run("setUp2", func(t *testing.T) { logSetUp(t, 2) }))

	assert.True(t, t.Run("sortByTimestampWithoutPattern", sortByTimestampWithoutPattern))
	assert.True(t, t.Run("sortByTimestampWithPattern", sortByTimestampWithPattern))
}


file/glob.go
// Package file implements methods for get logs from files.
//
// Log Finding.
package file

import (
	"fmt"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

const (
	splitAfter  = 2
	globPattern = "*-[0-9][0-9][0-9][0-9]*-*"
)

// GlobMatches map of Files
type GlobMatches struct {
	Files map[string]*FileMatches
}

// FileMatches map of Logs with same Category.
type FileMatches struct {
	Category string
	Logs     map[time.Time]*LogMatches
}

// LogMatches single line of log with timestamp.
type LogMatches struct {
	Year  int64
	Month int64
	Day   int64
	Time  time.Time
	Paths []string
}

func splitLogFilesByName(pattern string) *GlobMatches {
	var matches []string
	globMatches := &GlobMatches{
		Files: make(map[string]*FileMatches),
	}

	matches, _ = filepath.Glob(pattern)
	for _, match := range matches {
		matchWithoutExtension := strings.TrimSuffix(match, filepath.Ext(LogExtension))

		_, file := filepath.Split(matchWithoutExtension)
		title := strings.SplitN(file, "-", splitAfter)
		if len(title) < splitAfter {
			continue
		}

		dates := strings.SplitN(title[1], "-", splitAfter*2)
		if len(dates) < splitAfter {
			continue
		}

		fileMatches := &FileMatches{
			Category: title[0],
		}
		if _, ok := globMatches.Files[fileMatches.Category]; !ok {
			fileMatches.Logs = make(map[time.Time]*LogMatches)
			globMatches.Files[fileMatches.Category] = fileMatches
		}
		fileMatches = globMatches.Files[fileMatches.Category]

		logMatches := &LogMatches{}
		logMatches.Year, _ = strconv.ParseInt(dates[0], 10, 32)
		logMatches.Month, _ = strconv.ParseInt(dates[1], 10, 32)
		if len(dates) >= splitAfter+1 {
			logMatches.Day, _ = strconv.ParseInt(dates[2], 10, 32)
		}
		logMatches.Time = time.Date(
			int(logMatches.Year), time.Month(logMatches.Month), int(logMatches.Day),
			0, 0, 0, 0, time.UTC)
		if _, ok := fileMatches.Logs[logMatches.Time]; !ok {
			logMatches.Paths = make([]string, 0)
			fileMatches.Logs[logMatches.Time] = logMatches
		}
		logMatches = fileMatches.Logs[logMatches.Time]

		logMatches.Paths = append(logMatches.Paths, match)
	}

	return globMatches
}

// Glob search files into directory by pattern.
func Glob(dirPath string, pattern string, f func(matches *GlobMatches)) {
	// pay_debug-2017-11-15-ws02.log
	// pay_debug-2017-11-15-ws01.log
	// pay_debug-2017-11-15-ws03.log
	// pay_debug-2017-11-16-ws02.log
	// pay_debug-2017-11-16-ws01.log
	// pay_debug-2017-11-16-ws03.log
	// sms-2017-11-ws01.log
	// sms-2017-11-ws02.log
	// sms-2017-11-ws03.log
	// sms-2017-12-ws01.log
	// sms-2017-12-ws02.log
	// sms-2017-12-ws03.log

	// {"pay_debug": {"2017-11-15":["pay_debug-2017-11-15-ws01.log","pay_debug-2017-11-15-ws02.log"]}}

	if pattern == "" {
		pattern = fmt.Sprintf("%s%s", globPattern, LogExtension)
	}
	if dirPath != "" {
		pattern = fmt.Sprintf("%s/%s", filepath.Clean(dirPath), pattern)
	}

	f(splitLogFilesByName(pattern))
}


file/sort.go
// Package file implements methods for get logs from files.
//
// Log sorting.
package file

import (
	"sort"
)

// Aggregate extend Sort interface.
type Aggregate []Log

func (a Aggregate) Len() int {
	return len(a)
}

func (a Aggregate) Swap(i, j int) {
	a[i], a[j] = a[j], a[i]
}

func (a Aggregate) Less(i, j int) bool {
	return a[i].Timestamp.Unix() < a[j].Timestamp.Unix()

	//firstDate := a[i].Timestamp
	//secondDate := a[j].Timestamp
	//
	//firstDate.Unix()
	//
	//fmt.Println(firstDate)
	//fmt.Println(secondDate)
	//
	//return (firstDate.Year() < secondDate.Year()) &&
	//	(firstDate.Month() < secondDate.Month()) &&
	//	(firstDate.Day() < secondDate.Day()) &&
	//	(firstDate.Hour() < secondDate.Hour()) &&
	//	(firstDate.Minute() < secondDate.Minute()) &&
	//	(firstDate.Second() < secondDate.Second())
}

// SortByTimestamp sort log by timestamp.
func SortByTimestamp(logs []Log) []Log {
	sort.Sort(Aggregate(logs))
	return logs
}


cmd/colander/main.go
// Package main implements methods to colander log aggregator.
//
// Main.
package main

import (
	"fmt"
	"os"
	"strings"

	"github.com/supermercato24/colander/config"
	"github.com/urfave/cli"
)

const (
	ExIoerr = 74 // Exit Error
)

func init() {
	cli.VersionFlag = cli.BoolFlag{
		Name:  "version",
		Usage: "print the version number and exit",
	}
	cli.VersionPrinter = func(c *cli.Context) {
		fmt.Fprintf(c.App.Writer, "%s %s\n", strings.Title(config.Name), config.Version)
	}

	cli.HelpFlag = cli.BoolFlag{
		Name:  "help",
		Usage: "show this help message and exit",
	}
	cli.AppHelpTemplate = `usage: {{.HelpName}} {{if .VisibleFlags}}[optional options]{{end}}{{if .Commands}} command [command options]{{end}} {{if .ArgsUsage}}{{.ArgsUsage}}{{else}}[arguments...]{{end}}

{{.Name}} - {{.Usage}}{{if not .HideHelp}}{{ "\n" }}{{end}}
{{if .Commands}}positional arguments:
{{range .Commands}}{{if not .HideHelp}}  {{join .Names ", "}}{{ "\t" }}{{.Usage}}{{ "\n" }}{{end}}{{end}}{{end}}
{{if .VisibleFlags}}optional arguments:
  {{range .VisibleFlags}}{{.}}
  {{end}}
{{end}}`
}

func main() {
	var (
		dir     string
		pattern string
		remove  bool
		show    bool
	)

	app := cli.NewApp()
	app.Name = config.Name
	app.Version = config.Version
	app.Usage = "aggregate logs"
	//app.HideHelp = true
	//app.EnableBashCompletion = true
	app.Flags = []cli.Flag{
		cli.StringFlag{
			Name:        "d, dir",
			Usage:       "load logs from `DIR`",
			Destination: &dir,
		},
		cli.StringFlag{
			Name:        "p, pattern",
			Usage:       "load logs with this `PATTERN`",
			Destination: &pattern,
		},
		cli.BoolFlag{
			Name:        "r, remove",
			Usage:       "remove parsed logs",
			Destination: &remove,
		},
		cli.BoolFlag{
			Name:        "s, show",
			Usage:       "show output to screen",
			Destination: &show,
		},
	}
	app.CommandNotFound = func(c *cli.Context, command string) {
		cli.ShowAppHelp(c)

		fmt.Fprint(c.App.Writer, "command not found\n\n")
	}
	app.OnUsageError = func(c *cli.Context, err error, isSubcommand bool) error {
		if isSubcommand {
			return err
		}

		fmt.Fprint(c.App.Writer, "incorrect Usage\n\n")

		cli.ShowAppHelp(c)

		return nil
	}
	app.Action = func(c *cli.Context) error {

		if dir != "" {
			fileInfo, err := os.Stat(dir)
			if err != nil {
				return cli.NewExitError("input DIR doesn't not exist", ExIoerr)
			}
			mode := fileInfo.Mode()

			if exist := mode.IsDir(); !exist {
				return cli.NewExitError("input DIR is not a directory", ExIoerr)
			}
		} else if pattern == "" {
			return cli.NewExitError("input DIR is not a directory", ExIoerr)
		}

		colander(&colanderOptions{
			dir:     dir,
			pattern: pattern,
			remove:  remove,
			show:    show,
		})

		return nil
	}

	app.Run(os.Args)
}


cmd/colander/colander.go
// Package main implements methods to colander log aggregator.
//
// Command.
package main

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/supermercato24/colander/file"
)

type colanderOptions struct {
	dir     string
	pattern string
	remove  bool
	show    bool
}

func colander(options *colanderOptions) {
	var showClosure func(resultPath string, logs file.Aggregate)
	if options.show {
		showClosure = func(_ string, logs file.Aggregate) {
			for _, log := range logs {
				fmt.Print(string(log.Body))
			}
		}
	} else {
		showClosure = func(resultPath string, logs file.Aggregate) {
			err := file.LogWriteLines(resultPath, logs)
			if err != nil {
				panic(err)
			}
		}
	}

	var removeClosure func(resultPath string, paths []string)
	if options.remove {
		removeClosure = func(resultPath string, paths []string) {
			_, pathResult := filepath.Split(resultPath)
			for _, path := range paths {
				_, pathFile := filepath.Split(path)
				if pathResult != pathFile { // don't remove result file
					err := os.Remove(path)
					if err != nil {
						panic(err)
					}
				}
			}
		}
	} else {
		removeClosure = func(_ string, _ []string) {}
	}

	file.Glob(options.dir, options.pattern, func(matches *file.GlobMatches) {
		for _, match := range matches.Files {
			category := match.Category
			for _, match := range match.Logs {
				date := match.Time.Format("2006-01")

				logs := file.LogReadLines(match.Paths)
				file.SortByTimestamp(logs)

				var resultPath string
				if !options.show {
					resultPath = fmt.Sprintf("%s-%s", category, date)
					if match.Day != int64(0) {
						resultPath = fmt.Sprintf("%s-%d", resultPath, match.Time.Day())
					}
					resultPath = fmt.Sprintf("%s%s", resultPath, file.LogExtension)
					resultPath = filepath.Join(options.dir, resultPath)
				}

				showClosure(resultPath, logs)
				if len(match.Paths) > 1 {
					removeClosure(resultPath, match.Paths)
				}
			}
		}
	})
}


file/log.go
// Package file implements methods for get logs from files.
//
// Log structure.
package file

import (
	"bufio"
	"os"
	"regexp"
	"time"
)

const (
	LogExtension = ".log" // log extension
	NewLine      = '\n'   // log newline
)

// Log is a single transaction by timestamp.
type Log struct {
	Timestamp time.Time
	Body      []byte
}

var (
	timestampRegex = regexp.MustCompile(`[0-9]{4}-[0-9]{2}-[0-9]{2} [0-9]{2}:[0-9]{2}:[0-9]{2}`)
)

// Try to detect timestamp from body with regex and time parser.
// [2017-12-01 23:50:07]
// [2017-12-01 23:50:10]
func detectTimestamp(line []byte) (timestampDetected time.Time) {

	timestamp := timestampRegex.Find(line)
	if len(timestamp) > 0 {
		timestampDetected, _ = time.Parse("2006-01-02 15:04:05", string(timestamp))
	}

	return
}

// LogReadLines read lines of log file.
func LogReadLines(logFiles []string) []Log {
	var logs []Log

	for _, logFile := range logFiles {
		fd, err := os.Open(logFile)
		if err != nil {
			panic(err)
		}

		reader := bufio.NewReader(fd)

		for {
			line, err := reader.ReadBytes(NewLine)
			if err != nil {
				break
			}

			timestamp := detectTimestamp(line)
			if (time.Time{}) == timestamp {
				continue
			}
			log := Log{
				Timestamp: timestamp,
				Body:      line,
			}
			logs = append(logs, log)
		}
		fd.Close()
	}

	return logs
}

// LogWriteLines write lines into log file.
func LogWriteLines(path string, logs []Log) (err error) {
	os.Remove(path)
	fd, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return
	}
	defer fd.Close()

	writer := bufio.NewWriter(fd)

	for _, log := range logs {
		_, err = writer.Write(log.Body)
		if err != nil {
			return
		}
	}
	writer.Flush()

	return
}
